### PR TITLE
RobustParticleTransformer in `DataFormats/BTauReco`

### PR DESCRIPTION
**Pull request by AnnikaStein**
#### PR description:

A new jet flavour tagging algorithm (RobustParTAK4) is to be integrated soon, and relevant DataFormats need to be modified to handle more input variables (e.g. more features for charged / neutral candidates, as well as for secondary vertices). While we have a first full integration in our private recipes, we aim to integrate the respective producers and the inference part at a later time to implement improvements to the timing of the new modules with dynamic input sizes, which is a work-in-progress. This PR is for the data structure only, to meet the POG updates deadline.

Some relevant links to meetings where these developments have been discussed in BTV: https://indico.cern.ch/event/1218499/#5-adversarial-training-for-par and https://indico.cern.ch/event/1218501/#4-adversarial-training-for-par
And a short summary of the activities and milestones related to this project was part of the roundtable (https://indico.cern.ch/event/1236178/#4-general-status-and-feedback)

#### PR validation:

Basic set of code checks / code-format only (more dedicated tests will be added as part of the full integration, of which one can view a preview here: https://github.com/cms-sw/cmssw/compare/master...AnnikaStein:cmssw:from-CMSSW_13_0_0_pre3 where we tested the new inputs and outputs for the AOD to MiniAOD step, followed by mini to nano). More tests to follow in a not as time critical follow-up PR related to validation of inputs and outputs, which are however independent of the minimal changes introduced in this PR. With the present PR, we only create the necessary data structure (and will therefore not have to patch these in later, in line with https://twiki.cern.ch/twiki/bin/viewauth/CMS/ReleaseSchedule#Notes_on_Patch_Releases_formalis).

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Not a backport

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by AnnikaStein** [comment]
@perrotta 
I received a last minute change request - three features per candidate for the 4-vectors. Sorry for that.

**Comment by perrotta** [comment]
> @perrotta I received a last minute change request - three features per candidate for the 4-vectors. Sorry for that.

@AnnikaStein is your update complete by now?
Since, by chances, our entries in this thread almost overlapped, I think my "please build" actually already picked your last commit. Just to be sure. better to restart them nonetheless

**Comment by perrotta** [comment]
please abort

**Comment by AnnikaStein** [comment]
> > @perrotta I received a last minute change request - three features per candidate for the 4-vectors. Sorry for that.
> 
> @AnnikaStein is your update complete by now? Since, by chances, our entries in this thread almost overlapped, I think my "please build" actually already picked your last commit. Just to be sure. better to restart them nonetheless

Yes we must have typed our respective messages at the same time indeed.
This is the last update for this PR, we now have all the features that will be used ~now and in a future iteration, so from our side it's complete. Thanks a lot!

**Comment by perrotta** [comment]
please test

**Comment by mandrenguyen** [comment]
please test

**Comment by mandrenguyen** [comment]
type btv

**Comment by mandrenguyen** [comment]
+reconstruction
No changes to comparisons as expected.  
Updates data format in anticipation of upcoming flavor tagging developments.

File: DataFormats/BTauReco/src/classes_def.xml [hunk 1]
```diff
@@ -410,18 +414,24 @@
     <version ClassVersion="3" checksum="386672900"/>
     <version ClassVersion="4" checksum="113477759"/>
   </class>
-  <class name="btagbtvdeep::SecondaryVertexFeatures" ClassVersion="4">
+  <class name="btagbtvdeep::SecondaryVertexFeatures" ClassVersion="6">
+   <version ClassVersion="6" checksum="1869636758"/>
+   <version ClassVersion="5" checksum="1572623975"/>
```
Comment (perrotta) [inline DataFormats/BTauReco/src/classes_def.xml:419]: This intermediate "5" version is not needed -> move "6" to "5"

File: DataFormats/BTauReco/src/classes_def.xml [hunk 2]
```diff
@@ -410,18 +414,24 @@
...
    <version ClassVersion="4" checksum="3987593676"/>
     <version ClassVersion="3" checksum="3610780843"/>
   </class>
   <class name="btagbtvdeep::ShallowTagInfoFeatures" ClassVersion="3">
     <version ClassVersion="3" checksum="273372007"/>
   </class>
-  <class name="btagbtvdeep::NeutralCandidateFeatures" ClassVersion="4">
+  <class name="btagbtvdeep::NeutralCandidateFeatures" ClassVersion="6">
+   <version ClassVersion="6" checksum="267457446"/>
+   <version ClassVersion="5" checksum="2915967773"/>
```
Comment (perrotta) [inline DataFormats/BTauReco/src/classes_def.xml:428]: This intermediate "5" version is not needed -> move "6" to "5"

File: DataFormats/BTauReco/src/classes_def.xml [hunk 3]
```diff
@@ -410,18 +414,24 @@
...
    <version ClassVersion="4" checksum="3657612864"/>
     <version ClassVersion="3" checksum="938442516"/>
   </class>
-  <class name="btagbtvdeep::ChargedCandidateFeatures" ClassVersion="5">
+  <class name="btagbtvdeep::ChargedCandidateFeatures" ClassVersion="7">
+   <version ClassVersion="7" checksum="3294985591"/>
+   <version ClassVersion="6" checksum="2366727816"/>
```
Comment (perrotta) [inline DataFormats/BTauReco/src/classes_def.xml:434]: This intermediate "6" version is not needed -> move "7" to "6"

File: DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#ifndef DataFormats_BTauReco_ParticleTransformerAK4Features_h
+#define DataFormats_BTauReco_ParticleTransformerAK4Features_h
+
+//#include "DataFormats/BTauReco/interface/JetFeatures.h"
```
Comment (perrotta) [inline DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h:4]: Is this commented out code needed here? If not, please remove

File: DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#include "DataFormats/BTauReco/interface/NeutralCandidateFeatures.h"
+#include "DataFormats/BTauReco/interface/ChargedCandidateFeatures.h"
+
+#include <vector>
+
+namespace btagbtvdeep {
+
+  class ParticleTransformerAK4Features {
+  public:
+    //JetFeatures jet_features;
```
Comment (perrotta) [inline DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h:15]: Is this commented out code needed here? If not, please remove

File: DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    std::vector<SecondaryVertexFeatures> sv_features;
+
+    std::vector<NeutralCandidateFeatures> n_pf_features;
+    std::vector<ChargedCandidateFeatures> c_pf_features;
+
+    //std::size_t npv;  // used by deep flavour
```
Comment (perrotta) [inline DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h:22]: Is this commented out code needed here? If not, please remove

File: DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h (hunk 4 above)
Comment (AnnikaStein) [reply DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h:4]: Will remove the commented-out lines, thanks for pointing this out.
A bit more context: These global jet features currently only play a role for phase 2 developments, and have been used in the current DeepJet tagger, which is why these lines were included in the first place. As our current transformer model does not use them, I agree that these lines should be removed for now.

File: DataFormats/BTauReco/src/classes_def.xml (hunk 1 above)
Comment (AnnikaStein) [reply DataFormats/BTauReco/src/classes_def.xml:419]: Thanks for the comment, I was not sure about this when I applied the ClassVersions fix (as in, if one has to increment and keep track of all intermediate versions). I now went ahead with your suggestion to use the latest checksum and denote this as the new version "5".

With that fix applied, building works locally, as if there had never been an intermediate version.

File: DataFormats/BTauReco/src/classes_def.xml (hunk 2 above)
Comment (AnnikaStein) [reply DataFormats/BTauReco/src/classes_def.xml]: Same as for https://github.com/cms-sw/cmssw/pull/40706#discussion_r1099056341

And should be resolved with https://github.com/cms-sw/cmssw/pull/40706/commits/6727ef6e2ed76f0659c1b494adfeda38a53075e1

File: DataFormats/BTauReco/src/classes_def.xml (hunk 3 above)
Comment (AnnikaStein) [reply DataFormats/BTauReco/src/classes_def.xml]: Same as for https://github.com/cms-sw/cmssw/pull/40706#discussion_r1099056341

And should be resolved with https://github.com/cms-sw/cmssw/pull/40706/commits/6727ef6e2ed76f0659c1b494adfeda38a53075e1

File: DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h (hunk 5 above)
Comment (AnnikaStein) [reply DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h]: Same as for https://github.com/cms-sw/cmssw/pull/40706#discussion_r1099055626

And should be resolved with https://github.com/cms-sw/cmssw/pull/40706/commits/6727ef6e2ed76f0659c1b494adfeda38a53075e1

File: DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h (hunk 6 above)
Comment (AnnikaStein) [reply DataFormats/BTauReco/interface/ParticleTransformerAK4Features.h]: Same as for https://github.com/cms-sw/cmssw/pull/40706#discussion_r1099055626

And should be resolved with https://github.com/cms-sw/cmssw/pull/40706/commits/6727ef6e2ed76f0659c1b494adfeda38a53075e1

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1

**Comment by AnnikaStein** [comment]
Hi @nurfikri89 
as discussed via email, I support the modifications as they
- are first of all necessary to actually use the weights from `PuppiProducer` for a potential re-NANO, in the `packed_cand` case
- and second as this forces users to be explicit about what kind of weights they want to use.

Would we need a backport to 130X, given that one goal for this release reads „Possible re-miniAOD in EOY 2022“?

Best regards and thanks,
Annika

**Pull request merged**